keps/sig-auth/1688-dynamic-authentication-config/README.md

```diff
@@ -0,0 +1,620 @@
+and there is no concept of priority among them.  Token webhook implementations
+are responsible for understanding what tokens are meant for them.
+
+#### Source of identity
+
+One important characteristic of the an impersonation based proxy is that it is
+easy to trace when it asserts an identity by looking at the API server's audit
+logs.  This property will be maintained by setting audit annotations upon
+successful authentication.
+
+The `authenticationconfigs.authentication.k8s.io/name` key will be set to the
+`metadata.name` of the `AuthenticationConfig` object that asserted the `user.Info`.
+
+The `authenticationconfigs.authentication.k8s.io/type` key will be set to the
+`spec.type` of the `AuthenticationConfig` object that asserted the `user.Info`.
+
+More audit annotations may be added in the future to aid in tracing the source
+of the asserted identity.  For example, with the `oidc` type it may be helpful
+to include the issuer URL and client ID.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The proposed functionality will be gated behind a new feature flag called
+`DynamicAuthenticationConfig`.
+
+The proposed API will reside in the `authentication.k8s.io` group at version
+`v1alpha1` and resource `authenticationconfigs`.  The `AuthenticationConfig`
+`Kind` is defined as:
+
+```golang
+type AuthenticationConfig struct {
+  metav1.TypeMeta `json:",inline"`
+  // +optional
+  metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+  Spec AuthenticationConfigSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+}
+
+type AuthenticationConfigSpec struct {
+  Type AuthenticationConfigType `json:"type" protobuf:"bytes,1,opt,name=type"`
+
+  // this generic username and group prefix config is always required
+  // but the fields can be explicitly set to "-" to override it
+  Prefix *PrefixConfig `json:"prefix" protobuf:"bytes,2,opt,name=prefix"`
+
+  X509 *X509Config `json:"x509,omitempty" protobuf:"bytes,3,opt,name=x509"`
+
+  OIDC *OIDCConfig `json:"oidc,omitempty" protobuf:"bytes,4,opt,name=oidc"`
+
+  // This struct will be defined in a later revision
+  // Webhook *WebhookConfig `json:"webhook,omitempty" protobuf:"bytes,5,opt,name=webhook"`
+}
+
+type AuthenticationConfigType string
+
+const (
+  AuthenticationConfigTypeX509    AuthenticationConfigType = "x509"
+  AuthenticationConfigTypeOIDC    AuthenticationConfigType = "oidc"
+  AuthenticationConfigTypeWebhook AuthenticationConfigType = "webhook"
+)
+
+type X509Config struct {
+  // caBundle is a PEM encoded CA bundle used for client auth (x509.ExtKeyUsageClientAuth).
+  // +listType=atomic
+  // Required
+  CABundle []byte `json:"caBundle" protobuf:"bytes,1,opt,name=caBundle"`
```

Comment:
Please, *please* avoid the mistake of the aggregated apiserver api and make this `caBundle` a pointer to a separate Secret (or ConfigMap) and not inline.  The rest of the AuthenticationConfig is likely to be static (eg: helm) manifest, but the caBundle contents will be unique for every install.  This is _way_ easier to generate (and rotate) if it is a separate object (eg: you could just use cert-manager to maintain a self-signed CA).

Reply:
I agree: This turned out to be the most annoying part of deploying an admission Webhook.

I vaguely recall hearing the reason why that CA certificates was defined statically, but I can't find the discussion now in Slack. I think it had something to do with not wanting the API server to have to watch the _Secrets_ to reconfigure its Webhook client, though of course the API server already has to watch the Webhook configuration itself.

Reply:
I do not plan to diverge from the approach used by existing APIs here.  If they expand to allow references, I will mirror that here.  Overall I am not too worried about the `x509` case.  It is not really an extension point I expect to see used in helm charts.

For `webhook` I agree that this adds pain, but again, I do not plan on diverging from the existing APIs here.  Long term I would like Kube to provide first class support for a short-lived, auto rotated CA for the service network (and mint serving certs for all services that ask for one).  Then the common case of "I just want to run this webhook on the cluster please make the TLS nonsense go away" would be painless.

cc @munnerz since cert-manager is mentioned.

Reply:
To clarify, this is the config for the x509 authentication type (i.e. authenticating the user request), not for authenticating the apiserver request to the authenticator, right?

I believe the concerns raised here are with regards to the latter situation.

Reply:
> To clarify, this is the config for the x509 authentication type (i.e. authenticating the user request)

Correct.  This CA bundle allows validation of client certs.

> I believe the concerns raised here are with regards to the latter situation.

IIUC, the concern was "please let me separate my static config from dynamic config."  For example, when configuring an OPA admission webhook that enforces policy across all resources, all `ValidatingWebhookConfiguration` across all clusters will basically look the same minus the CA bundle.

One could imagine a similar situation where most of the `AuthenticationConfig` object is static minus the CA bundle.  I think it is far less likely and I must prefer being able to validate the CA bundle inline.  I think there could be a case for allowing an empty CA bundle which would be ignored until filled in by some controller.